curated_transformers/tokenization/tokenizer.py

```diff
@@ -0,0 +1,72 @@
+from typing import List, Type, TypeVar, Union
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+import torch
+
+from .._compat import has_hf_transformers, transformers
+
+# Only provided as typing.Self in Python 3.11+.
+Self = TypeVar("Self", bound="Tokenizer")
+
+
+@dataclass
+class PiecesWithIds:
+    pad_id: int
+    ids: List[List[int]]
+    lens: List[List[int]]
+    pieces: List[List[str]]
+
+    @property
+    def attention_mask(self):
+        """CPU tensor with attention masks.
+
+        The mask is equivalent to `ids.padded_tensor != padding_idx`."""
+        n_seqs = len(self.ids)
+        max_len = max(len(seq_ids) for seq_ids in self.ids)
+        mask = torch.full((n_seqs, max_len), False)
+        for idx, seq_ids in enumerate(self.ids):
+            mask[idx, : len(seq_ids)] = True
+        return mask
+
+    @property
+    def padded_tensor(self):
+        """Padded CPU tensor of the piece identifiers."""
+        n_seqs = len(self.ids)
+        max_len = max(len(seq_ids) for seq_ids in self.ids)
+        padded = torch.full((n_seqs, max_len), self.pad_id)
```

Comment:
I think this defaults to `torch.int64`. Given that 32-bits is plenty for the IDs, we might want to explicitly set `dtype=torch.int32` for better throughput on the GPU (when it eventually gets uploaded).

Reply:
Makes sense, I'll fix it (`int32` didn't work in the past for a lot of things like the embedding layer in the past, to I built up this habit).

Reply:
Fixed.